src/python_testing/TC_OPCREDS_3_9.py

```diff
@@ -0,0 +1,354 @@
+    CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BITS = 128
+    CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES = 16
+    CHIP_CRYPTO_AEAD_MIC_LENGTH_BITS = 128
+    CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES = 16
+    CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES = 13
+
+
+# From Matter spec src/crypto_primitives/crypto_primitives.py
+def bytes_from_hex(hex: str) -> bytes:
+    """Converts any `hex` string representation including `01:ab:cd` to bytes
+
+    Handles any whitespace including newlines, which are all stripped.
+    """
+    return unhexlify("".join(hex.replace(":", "").split()))
+
+
+# From Matter spec src/crypto_primitives/vid_verify_payload_test_vector.py
+# ECDSA-with-SHA256 using NIST P-256
+FABRIC_BINDING_VERSION_1 = 0x01
+STATEMENT_VERSION_1 = 0x21
+SIGNATURE_SIZE = (2 * MappingsV1.CHIP_CRYPTO_GROUP_SIZE_BYTES)
+SKID_SIZE = 20
+VID_VERIFICATION_CLIENT_CHALLENGE_SIZE_BYTES = 32
+ATTESTATION_CHALLENGE_SIZE_BYTES = MappingsV1.CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES
+
+
+def verify_signature(public_key: bytes, message: bytes, signature: bytes) -> bool:
+    """Verify a `signature` on the given `message` using `public_key`. Returns True on success."""
+
+    verifying_key: VerifyingKey = VerifyingKey.from_string(public_key, curve=NIST256p)
+    assert verifying_key.curve == NIST256p
+
+    try:
+        return verifying_key.verify(signature, message, hashfunc=hashlib.sha256)
+    except BadSignatureError:
+        return False
+
+
+# From Matter spec src/crypto_primitives/vid_verify_payload_test_vector.py
+def generate_vendor_fabric_binding_message(
+        root_public_key_bytes: bytes,
+        fabric_id: int,
+        vendor_id: int) -> bytes:
+
+    assert len(root_public_key_bytes) == MappingsV1.CHIP_CRYPTO_PUBLIC_KEY_SIZE_BYTES
+    assert fabric_id > 0 and fabric_id <= 0xFFFF_FFFF_FFFF_FFFF
+    assert vendor_id > 0 and vendor_id <= 0xFFF4
+
+    fabric_id_bytes = fabric_id.to_bytes(length=8, byteorder='big')
+    vendor_id_bytes = vendor_id.to_bytes(length=2, byteorder='big')
+    vendor_fabric_binding_message = FABRIC_BINDING_VERSION_1.to_bytes(
+        length=1) + root_public_key_bytes + fabric_id_bytes + vendor_id_bytes
+    return vendor_fabric_binding_message
+
+# From Matter spec src/crypto_primitives/vid_verify_payload_test_vector.py
+
+
+def generate_vendor_id_verification_tbs(fabric_binding_version: int,
+                                        attestation_challenge: bytes,
+                                        client_challenge: bytes,
+                                        fabric_index: int,
+                                        vendor_fabric_binding_message: bytes,
+                                        vid_verification_statement: Optional[bytes] = None) -> bytes:
+    assert len(attestation_challenge) == ATTESTATION_CHALLENGE_SIZE_BYTES
+    assert len(client_challenge) == VID_VERIFICATION_CLIENT_CHALLENGE_SIZE_BYTES
+    assert fabric_index > 0 and fabric_index < 255
```

Comment:
would it make sense to have a `VALID_FABRIC_INDEX_RANGE = range(1, 255)` and use it here and below in get_unassigned_fabric_index?

That could also contain a note why we say 255 is not a valid value and only 254 is ok.